File: sword_for_offer/codes/java/sfo_66_a_product_array_puzzle_s1/sfo_66_a_product_array_puzzle_s1.java
/*
* File: sfo_66_a_product_array_puzzle_s1.java
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

package sfo_66_a_product_array_puzzle_s1;

import include.*;
import java.util.*;

// ===== Solution Code =====
class Solution {
    public int[] constructArr(int[] a) {
        int len = a.length;
        if(len == 0) return new int[0];
        int[] b = new int[len];
        b[0] = 1;
        int tmp = 1;
        for(int i = 1; i < len; i++) {
            b[i] = b[i - 1] * a[i - 1];
        }
        for(int i = len - 2; i >= 0; i--) {
            tmp *= a[i + 1];
            b[i] *= tmp;
        }
        return b;
    }
}

public class sfo_66_a_product_array_puzzle_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        int[] a = { 1, 2, 3, 4, 5 };
        // ====== Driver Code ======
        Solution slt = new Solution();
        int[] res = slt.constructArr(a);
        System.out.println(Arrays.toString(res));
    }
}

File: sword_for_offer/codes/python/sfo_66_a_product_array_puzzle_s1.py
'''
File: sfo_66_a_product_array_puzzle_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def constructArr(self, a: List[int]) -> List[int]:
        b, tmp = [1] * len(a), 1
        for i in range(1, len(a)):
            b[i] = b[i - 1] * a[i - 1] # 下三角
        for i in range(len(a) - 2, -1, -1):
            tmp *= a[i + 1]            # 上三角
            b[i] *= tmp                # 下三角 * 上三角
        return b

# ======= Test Case =======
a = [1, 2, 3, 4, 5]
# ====== Driver Code ======
slt = Solution()
res = slt.constructArr(a)
print(res)
